### DISCUSSION TITLE
Unable to connect to WIFI with S2-mini clone

### DISCUSSION THREAD

**Discussion by Tico06** [micropython/micropython#12390] [ESP32]
Hi guys,

I am trying to connect to WIFI with a S2-mini clone. The vendor says it's ESP32-S2FN4R2.

I have compile the following firmware:
MicroPython v1.20.0-450-g3637252b7-dirty on 2023-09-07; LOLIN_S2_MINI with ESP32-S2FN4R2

Here is my code in Thonny:
```
>>> import network
>>>  wlan = network.WLAN(network.STA_IF)
>>>  wlan.active(True)
True
>>> wlan.scan()
[(b'ssid', b'\xea2\xdaA_\xc8', 8, -57, 2, False)]
>>> wlan.connect('ssid', 'pwd')
>>> wlan.isconnected()
False
>>> wlan.status() == network.STAT_CONNECTING
True
>>> wlan.status() == network.STAT_CONNECTING
True
>>> wlan.status() == network.STAT_WRONG_PASSWORD
False
>>> 

```

SSID and PWD are validated with my laptop.

I have no idea where to look.

Thanks and regards,

Eric.

**Comment by davefes** [comment]
After `wlan.active(True)` try:
```py
#  seems more reliable to start with a fresh connect()
    if wlan.isconnected():
        wlan.disconnect()
        print (f'started in the connected state, but now disconnected')
    else:
        print (f'started in the disconnected state')
```
and after any `wlan.disconnect()` place:
```py
    wlan.active(False)
```

**Reply by Tico06** [reply]
Hi There,

Thanks a lot for the quick reply.

Unfortunately, didn't fix the issue.

regards,

Eric.

**Comment by davefes** [comment]
You did print out `'pwd'` to verify that it is correct?

**Reply by Tico06** [reply]
yes, double checked.

**Comment by robert-hh** [comment]
Which frequency & channel is used by your AP?

**Reply by Tico06** [reply]
Hi Robert !

2.447Ghz on channel 8

Thx and reg

Eric.

**Comment by steveh127** [comment]
Whenever I'm making a network connection, whatever the chip, I put the connect code in a while loop to try multiple times in succession. I also trap and ignore any errors in the loop though you don't seem to get those.  It often takes a number of tries to connect.

**Comment by Tico06** [comment]
Hi All,

Thanks Steve. Here is the code I wrote to test:
```
import network

wlan = network.WLAN(network.STA_IF) # create station interface
wlan.active(True)       # activate the interface
if wlan.isconnected():
    wlan.disconnect()
    print (f'Started in the connected state, but now disconnected.')
else:
    print (f'Started in the disconnected state.')

wlan.config(reconnects=0)

while (not wlan.isconnected()):
    print (wlan.scan())             # scan for access points
    print (f'Connecting', accessPoint)
    wlan.connect(accessPoint, apPWD) # connect to an AP
    while (wlan.status() == network.STAT_CONNECTING):
        pass
    if (wlan.status() == network.STAT_IDLE):
        print (f'STAT_IDLE:', wlan.status())
    elif (wlan.status() == network.STAT_CONNECTING):
        print (f'STAT_CONNECTING:', wlan.status())
    elif (wlan.status() == network.STAT_WRONG_PASSWORD):
        print (f'STAT_WRONG_PASSWORD:', wlan.status())
    elif (wlan.status() == network.STAT_NO_AP_FOUND):
        print (f'STAT_NO_AP_FOUND:', wlan.status())
#    elif (wlan.status() == network.STAT_CONNECT_FAIL):
#        print (f'STAT_CONNECT_FAIL', wlan.status())
    elif (wlan.status() == network.STAT_GOT_IP):
        print (f'STAT_GOT_IP:', wlan.status())
    else:
        print(f'Unknown status:', wlan.status())

print (f'MAC:', wlan.config('mac'))      # get the interface's MAC address
print (f'ifconfig:', wlan.ifconfig())         # get the interface's IP/netmask/gw/DNS addresses
```

I have two access points I should be able to connect to. My internet box and I setup an access point with my mobile to test. Here are the results of the code above trying accessing these access points:

```
Started in the connected state, but now disconnected.
[(b'mobile_ap', b'>\xde\xceslP', 1, -40, 3, False), (b'box_ap', b'\xea2\xdaA_\xc8', 8, -68, 2, False)]
Connecting box_ap
STAT_NO_AP_FOUND: 201
[(b'mobile_ap', b'>\xde\xceslP', 1, -40, 3, False), (b'box_ap', b'\xea2\xdaA_\xc8', 8, -68, 2, False)]
Connecting box_ap
STAT_NO_AP_FOUND: 201
[(b'mobile_ap', b'>\xde\xceslP', 1, -40, 3, False), (b'box_ap', b'\xea2\xdaA_\xc8', 8, -68, 2, False)]
Connecting box_ap
STAT_NO_AP_FOUND: 201
[(b'mobile_ap', b'>\xde\xceslP', 1, -40, 3, False), (b'box_ap', b'\xea2\xdaA_\xc8', 8, -68, 2, False)]
Connecting box_ap
```

Here, box_ap is not found on the radio with the wlan.connect(accessPoint, apPWD), but it's listed from the wlan.scan().

Now connecting to my mobile ap:
```
Started in the connected state, but now disconnected.
[(b'mobile_ap', b'>\xde\xceslP', 1, -40, 3, False), (b'box_ap', b'\xea2\xdaA_\xc8', 8, -68, 2, False)]
Connecting mobile_ap
STAT_GOT_IP: 1010
MAC: b'\xc0N0U\xbfR'
ifconfig: ('192.168.99.164', '255.255.255.0', '192.168.99.184', '192.168.99.184')
```
As you can see, it's connected straight.

Unfortunately, I  need to connect to my box access point. Is it possible to configure the phy_mode in ESP32? I'll give it a try.

Thanks and regards,

Eric.

**Comment by Tico06** [comment]
Unfortunatelly, py_mode doesn't look implemented for ESP32:
```
>>> network.phy_mode(network.MODE_11B)
>>> print (network.phy_mode())
None
```

**Comment by shariltumin** [comment]
This may be a silly question. I'm wondering if you can connect your smartphone to the 'box_ap'. That should probably be OK.

Trying with my ESP32-S2FN4R2 boards, one as AP, the other as STA, I have never seen status 201 (STAT_NO_AP_FOUND).

Attempt to connect to non-existent AP, I get status 1001 (STAT_CONNECTING) until I give up.

This is my 'wifi_connect.py' script

```py
import time, network
network.WLAN(network.AP_IF).active(False)
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.disconnect()
ssid, pwd = "ABC", "abc123def"
wlan.connect(ssid, pwd)

cnt = 0
while True:
    stat = wlan.status()
    if stat == network.STAT_GOT_IP:
       print('Connected with IP:', wlan.ifconfig()[0])
       break
    else:
       if stat == network.STAT_CONNECTING:
          # print('Connecting ... status', stat)
          time.sleep_ms(500)
          cnt += 1
          if cnt>10:
             print('Give up, trying to connect', cnt, 'times')
             break
       else:
          print('Give up with status', stat)
          break

```

This is my 'AP.py' script

```py
import network, time

SID = 'ABC'
PWD = 'abc123def'
IP = '10.10.4.1'

network.WLAN(network.STA_IF).active(False)
time.sleep(1) # wait for IF to reset
ap = network.WLAN(network.AP_IF)
ap.active(True)
ap.config(essid=SID)
ap.config(authmode=3, password=PWD)
# IP address, netmask, gateway, DNS
z = (IP, '255.255.255.0', IP, '8.8.8.8')
ap.ifconfig(z)
print(ap.ifconfig())

```

**Reply by Tico06** [reply]
Hi Shariltumin,

Thanks a lot for stepping in.

> This may be a silly question. I'm wondering if you can connect your smartphone to the 'box_ap'. That should probably be OK.

box_ap is my home internet box, and I am currently using it to feed this thread. So yes, box_ap works fine. And I copy/pasted SSID and PWD from my settings on my laptop to micropython code. But maybe you had an idea behing connecting my smartphone to it beyond validating ? Anyway, I double checked with my mobile and it connected and browsed just fine.

>Trying with my ESP32-S2FN4R2 boards, one as AP, the other as STA, I have never seen status 201 (STAT_NO_AP_FOUND).
>
>Attempt to connect to non-existent AP, I get status 1001 (STAT_CONNECTING) until I give up.

I think I could explain this. By default, the driver retries indefinitevly to connect to WIFI. So at the time it fails, it retries straight. I guess this is why you never see other status than STAT_CONNECTING. If you look at my code, I have a wlan.config(reconnects=0) which prevents reconnect attempts. In my loop, the reconnect attempts are manual with wlan.connect(accessPoint, apPWD) , once it stop connecting, tested with the empty while loop, I check the status, and I loop until isconnected() returns True.
With my code when I attempt to connect to a non-existing AP, I have  STAT_NO_AP_FOUND: 201 returned. And I have same status when trying to connect to my box_ap. I'll give it a try on a different channel.

What I can see from the scan, I have three AP on channel 8 created by my box:
`(b'box_ap', b'\xea2\xdaA_\xc8', 8, -48, 2, False), (b'', b'\xea2\xdaA_\xc9', 8, -48, 3, False), (b'box_ap_secure', b'\xea2\xdaA_\xca', 8, -49, 5, False)`
Could it disturb ?

Thanks and regards,

Eric.

**Reply by shariltumin** [reply]
Or maybe your box_ap is running at 5Ghz?

**Reply by Tico06** [reply]
nop, 2.4Ghz

**Comment by Tico06** [comment]
Hi All,

I tried to connect with BSSID, resulting with same status. I changed the channel, same result.... Stuck here.

**Comment by shariltumin** [comment]
There is one more thing you can try. You can connect your smartphone to your 'box_ap' and enable Wi-Fi Hotspot on your phone. Then try to connect your esp32-s2 to your smartphone.

I did the same test here with my old Android phone and the wemos s2 mini card. I managed to connect and get an IP.

**Reply by shariltumin** [reply]
No, it is not possible to use my phone as a hotspot while connected to Wi-Fi

**Comment by robert-hh** [comment]
Since the early times of the ESP8266 I use the below script for the connection to the AP. It is derived from the example script in the documentation. connect() is only called once, and then it just checks the state using is_connected(). I verified that it works with a ESP32S2 board (UM Feather S2).
```
WIFI_SSID = "<ssid>"
WIFI_PASSWD = "<passwd>"

def do_connect():
    import network
    import time

    wlan = network.WLAN(network.STA_IF) # create station interface
    wlan.active(True)       # activate the interface
    if not wlan.isconnected():      # check if the station is connected to an AP
        wlan.connect(WIFI_SSID, WIFI_PASSWD) # connect to an AP

        for _ in range(20):
            if wlan.isconnected():      # check if the station is connected to an AP
                break
            print('.', end='')
            time.sleep(0.5)
        else:
            print(" Connect attempt timed out\n")
            return
    print('\nnetwork config:', wlan.ifconfig())

do_connect()
```

**Reply by Tico06** [reply]
Hi Robert,

Thanks a lot for your contribution. Your code works to connect to my mobile:```
```
>>> do_connect()
...................
network config: ('192.168.110.164', '255.255.255.0', '192.168.110.243', '192.168.110.243')
```

But doesn't work to connect to my box. Thus no progress:
```
>>> do_connect()
.................... Connect attempt timed out
```

From my point of view, this is not a code issue, but a compatibility issue between micropython or the S2 and my wifi box. I'll have a look (again) in my box configuration and see what I can change here.

thanks and regard,

Eric.

**Comment by shariltumin** [comment]
This is not related to the problem you are experiencing, but installing a NAT router may somewhat solve the problem of connecting your esp32-s2 board directly to your main Internet router.

YouTube: https://www.youtube.com/watch?v=41Lymi6rXA8

URL: https://github.com/martin-ger/esp32_nat_router

I have installed this and am using the network it provides for my IoT experiments.

**Reply by Tico06** [reply]
Hi Shariltumin,

Well... I guess you are right, but having a NAT router of 20-50€ to connect a 4€ device to internet is a bit strange to me....
I'll have a look in the ESP32 nat router, looks interesting.

Anyway, it should work right and easier...

Thanks again for your suggestions. For the time being I'll keep working on my development connecting to Internet with my mobile. And when needed, I'll spend more time on sorting out with my Internet box and investigate further on NAT.

Regards,

Eric.
